### PR TITLE
build: enable release builds on Windows

### PR DESCRIPTION
**Pull request by sesmith177**
*Description*:

Now that #5218 has been merged, we can enable release builds on Windows. This PR updates `ci/do_ci.ps1` to build release + fastbuild versions in addition to debug versions. It also removes support for build scripts in `ci/build_container/build_recipes/` on Windows, as there is no clean way to pass information from Bazel regarding the type of build to those scripts

*Risk Level*:
Low

*Testing*:
`bazel build //source/... && bazel test //test/...`

*Docs Changes*:
N/A
*Release Notes*:
N/A

### CODE REVIEW

File: bazel/foreign_cc/BUILD [hunk 1]
```diff
@@ -13,12 +42,46 @@ cmake_external(
...
+        ":windows_dbg_build": "cp $BUILD_TMPDIR/CMakeFiles/c-ares.dir/c-ares.pdb $INSTALLDIR/lib/c-ares.pdb",
+        "//conditions:default": "",
+    }),
+    static_libraries = select({
+        ":windows_x86_64": ["cares.lib"],
+        "//conditions:default": ["libcares.a"],
+    }),
+)
+
+cmake_external(
```
Comment (htuch) [inline bazel/foreign_cc/BUILD]: Can we create an `envoy_cmake_external` macro that hides the boiler plate?

File: bazel/foreign_cc/nghttp2.patch [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+   # Set it to "int" to match the behavior of AC_TYPE_SSIZE_T (autotools).
+-  set(ssize_t int)
++  if(WIN32 AND CMAKE_SIZEOF_VOID_P EQUAL 8)
++    set(ssize_t int64_t)
++  else()
++    set(ssize_t int)
++  endif()
+ endif()
+ # AC_TYPE_UINT8_T
+ # AC_TYPE_UINT16_T
```
Comment (htuch) [inline bazel/foreign_cc/nghttp2.patch]: What are the plans to upstream this? As it stands, having this patch will make version bumps hard as we will need to treat the patch manually.

File: bazel/foreign_cc/BUILD (hunk 1 above)
Comment (lizan) [reply bazel/foreign_cc/BUILD]: The real boiler plate part is `generate_crosstool_file`, why this is needed @sesmith177? If it is a general issue probably upstream to `rules_foreign_cc`?

Comment (sesmith177) [reply bazel/foreign_cc/BUILD]: According to the README (and our testing), you need to explicitly specify the option to generate the CMake crosstool file for `rules_foreign_cc` to work on Windows

File: bazel/foreign_cc/nghttp2.patch (hunk 2 above)
Comment (sesmith177) [reply bazel/foreign_cc/nghttp2.patch]: A previous attempt to fix the `ssize_t` issue on Windows: https://github.com/nghttp2/nghttp2/issues/616 was reverted, since adding the typedef could break projects that have code that also `#define ssize_t` on Windows.

For the rest of the patch, there's been no movement on the PR we submitted back in July

File: bazel/foreign_cc/BUILD (hunk 1 above)
Comment (htuch) [reply bazel/foreign_cc/BUILD]: The boilerplate to me is both the `postfix_script` library retrieval of PDB and the crosstool code. Is there a better way to support them in `rules_foreign_cc`? CC @irengrig.

We also have the ninja configuration as boilerplate. Hence why I reckon it would be useful to make a `envoy_cmake_external` and hide some of this away, but I agree with @lizan we could push further back into `rules_foreign_cc` some more standard things.

File: bazel/foreign_cc/nghttp2.patch (hunk 2 above)
Comment (htuch) [reply bazel/foreign_cc/nghttp2.patch]: Thanks for the update. It would be great if we can get this Windows support landed upstream, CC @tatsuhiro-t. Do we at least have a roadmap for how to do this?

**Comment by sesmith177** [comment]
@htuch I've added a macro that hides at least some of the boilerplate (and also handles removing CMAKE_BUILD_TYPE on Windows). I think it should be possible to push the PDB selection in there as well, though handling the fact that `benchmark` has its own `postfix_script` might be a little tricky

File: bazel/envoy_build_system.bzl [hunk 3]
```diff
@@ -195,6 +196,40 @@ def envoy_include_prefix(path):
...
+        make_commands = ["ninja", "ninja install"],
+        lib_source = "",
+        postfix_script = "",
+        static_libraries = []):
+    # On Windows, we don't want to explicitly set CMAKE_BUILD_TYPE,
+    # rules_foreign_cc will figure it out for us
+    cache_entries_no_build_type = {}
+    for key in cache_entries.keys():
+        if key != "CMAKE_BUILD_TYPE":
+            cache_entries_no_build_type[key] = cache_entries[key]
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Does Starlark not have dictionary comprehensions? Should be 1 line in Python.

File: bazel/foreign_cc/BUILD [hunk 4]
```diff
@@ -1,85 +1,108 @@
...
     },
-    cmake_options = ["-GNinja"],
     lib_source = "@com_github_c_ares_c_ares//:all",
-    make_commands = [
-        "ninja",
-        "ninja install",
-    ],
-    static_libraries = ["libcares.a"],
+    postfix_script = select({
+        "//bazel:windows_dbg_build": "cp $BUILD_TMPDIR/CMakeFiles/c-ares.dir/c-ares.pdb $INSTALLDIR/lib/c-ares.pdb",
```
Comment (htuch) [inline bazel/foreign_cc/BUILD]: Yeah, to the extent you can push in most of this that make sense and parameterize by path it'd be good. We can still add a postfix script for additional stuff as needed on each target.

File: bazel/foreign_cc/nghttp2.patch (hunk 2 above)
Comment (tatsuhiro-t) [reply bazel/foreign_cc/nghttp2.patch]: How about adding a cmake flag to define ssize_t as int64_t?  By default, ssize_t is int for backward compatibility.

**Comment by sesmith177** [comment]
@htuch ready for another look

File: bazel/envoy_build_system.bzl [hunk 5]
```diff
@@ -195,6 +196,40 @@ def envoy_include_prefix(path):
...
+        cmake_options = ["-GNinja"],
+        make_commands = ["ninja", "ninja install"],
+        lib_source = "",
+        postfix_script = "",
+        static_libraries = []):
+    # On Windows, we don't want to explicitly set CMAKE_BUILD_TYPE,
+    # rules_foreign_cc will figure it out for us
+    cache_entries_no_build_type = {}
+    for key in cache_entries.keys():
+        if key != "CMAKE_BUILD_TYPE":
```
Comment (irengrig) [inline bazel/envoy_build_system.bzl]: CMAKE_BUILD_TYPE is calculated by cmake_external internally, from the type of Bazel build.
I added the way how to avoid that: https://github.com/bazelbuild/rules_foreign_cc/pull/224

Please pass "CMAKE_BUILD_TYPE": "" (empty string) to cmake_external, and then the computed value for CMAKE_BUILD_TYPE will not be used (there would be no "-DCMAKE_BUILD_TYPE=<something>" on the cmake call.)

Comment (sesmith177) [reply bazel/envoy_build_system.bzl]: On Windows, we want to use the computed value for CMAKE_BUILD_TYPE so we can get the desired release / debug builds

File: bazel/envoy_build_system.bzl [hunk 6]
```diff
@@ -195,6 +196,57 @@ def envoy_include_prefix(path):
...
+        static_libraries = [],
+        copy_pdb = False):
+    # On Windows, we don't want to explicitly set CMAKE_BUILD_TYPE,
+    # rules_foreign_cc will figure it out for us
+    cache_entries_no_build_type = {key: cache_entries[key] for key in cache_entries.keys() if key != "CMAKE_BUILD_TYPE"}
+
+    pf = ""
+    if copy_pdb:
+        name_to_pdb = {"ares": "c-ares", "nghttp2": "nghttp2_static", "zlib": "zlibstatic"}
+        pdb_name = name_to_pdb.get(name, name)
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: I like all of this PR except this bit :) Can we pass in the pdf_name as an arg to `envoy_cmake_external`, to avoid having target specific logic/defines in the macro?

Comment (sesmith177) [reply bazel/envoy_build_system.bzl]: makes sense, done

File: bazel/foreign_cc/BUILD [hunk 7]
```diff
@@ -1,85 +1,101 @@
...
-    static_libraries = ["libcares.a"],
+    pdb_name = "c-ares",
+    static_libraries = select({
+        "//bazel:windows_x86_64": ["cares.lib"],
+        "//conditions:default": ["libcares.a"],
+    }),
 )
 
-cmake_external(
+envoy_cmake_external(
```
Comment (lizan) [inline bazel/foreign_cc/BUILD]: Hmm, I missed this, why did you need to add this? repositories.bzl doesn't add the `:all` target to benchmark and it is built with upstream `BUILD.bazel` file. #5890

Comment (htuch) [reply bazel/foreign_cc/BUILD]: Yeah, @sesmith177 could you followup on this one? Thanks.

Comment (sesmith177) [reply bazel/foreign_cc/BUILD]: Yeah, that's definitely a mistake -- I hadn't noticed that when https://github.com/envoyproxy/envoy/commit/7fa55135828dc8d93c6eee6f363534afa5cb3484 removed the `benchmark.sh` script, it did not add a corresponding `cmake_external` rule. So when I merged our branch in to fix the Windows issues, this extra rule came with it.

Sorry for the confusion.
